### PR TITLE
IOTPROT-663 simplify DatagramSocketTransportAPI

### PR DESCRIPTION
**Pull request by samhaa01**

### CODE REVIEW

**Review by szysas — APPROVED** [review]
LGTW, can you squash commits?

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 1]
```diff
@@ -134,39 +139,65 @@ protected void receive(CoapReceiver coapReceiver, DatagramPacket datagramPacket)
...
+    }
+
+    protected void sendPacket(DatagramPacket packet) throws IOException {
+        this.socket.send(packet);
+    }
+
+    protected void createSocket() throws SocketException {
+        socket = new RealDatagramSocket(bindAddress);
+        socket.setSoTimeout(200);
     }
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:189]: Looks like those new methods are not used anywhere.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 2]
```diff
@@ -72,30 +76,32 @@ public DatagramSocketTransport(int localPort) {
...
-            socket = createSocket();
+        if (!socketCreated()) {
+            createSocket();
         }
 
         if (socketBufferSize > 0) {
-            socket.setReceiveBufferSize(socketBufferSize);
-            socket.setSendBufferSize(socketBufferSize);
+            updateReceiveBufferSize();
+            updateSendBufferSize();
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:101]: Lets remove that from here. Setting buffer sizes could be done in actual implementation of datgramsocket.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+package com.mbed.coap.transport.udp;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+
+public interface IDatagramSocket {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java:23]: I'd recommend to rename into `BlockingSocket`

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    void send(DatagramPacket p) throws IOException;
+
+    void receive(DatagramPacket p) throws IOException;
+
+    void connect(SocketAddress addr) throws SocketException;
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java:28]: This looks like is not in use.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    void setSendBufferSize(int size) throws SocketException;
+
+    void setReceiveBufferSize(int size) throws SocketException;
+
+    int getSendBufferSize() throws SocketException;
+
+    int getReceiveBufferSize() throws SocketException;
+
+    boolean getReuseAddress() throws SocketException;
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java:52]: These could be done in actual implementation. No need to expose in this interface.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java (hunk 3 above)
Comment (szysas) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java:23]: This interface could be quite simple, for example:
```
public interface BlockingSocket extends Closeable {

    void send(DatagramPacket datagramPacket) throws IOException;

    void receive(DatagramPacket datagramPacket) throws IOException;

    InetSocketAddress getBoundAddress();

}
```

File: coap-core/src/main/java/com/mbed/coap/transport/udp/RealDatagramSocket.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+package com.mbed.coap.transport.udp;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketException;
+
+public class RealDatagramSocket implements IDatagramSocket {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/RealDatagramSocket.java:25]: This could be also simplified:
```suggestion
public class DatagramSocketAdapter extends DatagramSocket implements BlockingSocket {

    public DatagramSocketAdapter(InetSocketAddress bindSocket) throws SocketException {
        super(bindSocket);
        setReuseAddress(false);
    }

    public DatagramSocketAdapter(int port) throws SocketException {
        super(port);
        setReuseAddress(false);
    }

    @Override
    public InetSocketAddress getBoundAddress() {
        return ((InetSocketAddress) getLocalSocketAddress());
    }

}
```

File: coap-core/src/main/java/com/mbed/coap/transport/udp/BlockingSocket.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mbed.coap.transport.udp;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+
+public interface BlockingSocket {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/BlockingSocket.java:21]: Use Closeable
```suggestion
public interface BlockingSocket extends Closeable {
```

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 8]
```diff
@@ -160,13 +159,16 @@ public void sendPacket0(CoapPacket coapPacket, InetSocketAddress adr, TransportC
...
+    }
+
     public InetSocketAddress getLocalSocketAddress() {
-        return (InetSocketAddress) socket.getLocalSocketAddress();
+        return (InetSocketAddress) getSocket().getLocalSocketAddress();
     }
 
-    protected DatagramSocket getSocket() {
-        return socket;
+    public DatagramSocketAdapter getSocket() {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:170]: Change return type to `BlockingSocket`, it will not always be `DatagramSocketAdapter`.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 9]
```diff
@@ -72,33 +71,34 @@ public DatagramSocketTransport(int localPort) {
...
 
         if (socketBufferSize > 0) {
-            socket.setReceiveBufferSize(socketBufferSize);
-            socket.setSendBufferSize(socketBufferSize);
+            getSocket().setReceiveBufferSize(socketBufferSize);
+            getSocket().setSendBufferSize(socketBufferSize);
         }
-        socket.setReuseAddress(reuseAddress);
-        LOGGER.info("CoAP server binds on " + socket.getLocalSocketAddress());
+        getSocket().setReuseAddress(reuseAddress);
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:98]: This can be removed

**Review by szysas — CHANGES_REQUESTED** [review]

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java [hunk 10]
```diff
@@ -66,61 +65,48 @@ public void initializingWithStateException() throws IOException {
...
-        trans.setSocketBufferSize(12345);
+        DatagramSocketAdapter datagramSocket = new DatagramSocketAdapter(new InetSocketAddress(0), false, 12345);
+        DatagramSocketTransport trans = new DatagramSocketTransport(datagramSocket, mock(Executor.class));
         trans.start(mock(CoapReceiver.class));
 
-        assertTrue(trans.getSocket().isBound());
-        assertFalse(trans.getSocket().isClosed());
-        assertEquals(12345, trans.getSocket().getSendBufferSize());
-        assertFalse(trans.getSocket().getReuseAddress());
+        assertTrue(((DatagramSocketAdapter) trans.getSocket()).isBound());
```
Comment (szysas) [inline coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java:79]: This is the same as `datagramSocket`, so why to cast?

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 8 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/BlockingSocket.java (hunk 7 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/BlockingSocket.java]: fixed.

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java (hunk 10 above)
Comment (samhaa01) [reply coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java (hunk 5 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java (hunk 4 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java (hunk 3 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/IDatagramSocket.java]: done.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 1 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package com.mbed.coap.transport.udp;
+
+import java.net.DatagramSocket;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+
+public class DatagramSocketAdapter extends DatagramSocket implements BlockingSocket {
+    protected int socketBufferSize = -1;
+    boolean reuseAddress;
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java:24]: Why is this needed?

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 12]
```diff
@@ -160,13 +135,16 @@ public void sendPacket0(CoapPacket coapPacket, InetSocketAddress adr, TransportC
...
+
     public InetSocketAddress getLocalSocketAddress() {
-        return (InetSocketAddress) socket.getLocalSocketAddress();
+        return (InetSocketAddress) ((DatagramSocketAdapter) getSocket()).getLocalSocketAddress();
     }
 
-    protected DatagramSocket getSocket() {
-        return socket;
+    public BlockingSocket getSocket() {
+        return (DatagramSocketAdapter) socket;
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:147]: Why casting here?

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 13]
```diff
@@ -160,13 +135,16 @@ public void sendPacket0(CoapPacket coapPacket, InetSocketAddress adr, TransportC
...
     }
 
-    @Override
+    protected boolean socketCreated() {
+        return (socket != null);
+    }
+
     public InetSocketAddress getLocalSocketAddress() {
-        return (InetSocketAddress) socket.getLocalSocketAddress();
+        return (InetSocketAddress) ((DatagramSocketAdapter) getSocket()).getLocalSocketAddress();
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:143]: How this will work when socket is other type than `DatagramSocketAdapter`?

File: coap-core/src/test/java/com/mbed/coap/server/internal/CoapTcpMessagingTest.java [hunk 14]
```diff
@@ -17,13 +17,9 @@
 
 import static org.assertj.core.api.Assertions.*;
 import static org.junit.Assert.*;
-import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.*;
```
Comment (szysas) [inline coap-core/src/test/java/com/mbed/coap/server/internal/CoapTcpMessagingTest.java:20]: I guess not needed change.

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java [hunk 15]
```diff
@@ -66,61 +65,48 @@ public void initializingWithStateException() throws IOException {
...
-            } catch (Exception e) {
-                assertTrue(e instanceof IllegalStateException);
-            }
         } finally {
             trans.stop();
         }
     }
 
     @Test
     public void initializeWithParameters() throws Exception {
```
Comment (szysas) [inline coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java:74]: I think this test does not make sense any more, just remove it.

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java [hunk 16]
```diff
@@ -66,61 +65,48 @@ public void initializingWithStateException() throws IOException {
...
+        assertEquals(12345, datagramSocket.getSendBufferSize());
+        assertFalse(datagramSocket.getReuseAddress());
 
         trans.stop();
-        assertTrue(trans.getSocket().isClosed());
+        assertTrue(datagramSocket.isClosed());
     }
 
     @Test
     public void reopenSamePort() throws IOException {
```
Comment (szysas) [inline coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java:89]: I think this test does not make sense any more, just remove it.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 9 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: Removed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java (hunk 11 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java]: ok. I removed that. I don't know why it was there.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+ * limitations under the License.
+ */
+package com.mbed.coap.transport.udp;
+
+import java.net.DatagramSocket;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+
+public class DatagramSocketAdapter extends DatagramSocket implements BlockingSocket {
+    protected int socketBufferSize = -1;
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java:23]: What is this needed for?

File: coap-core/src/test/java/com/mbed/coap/server/internal/CoapTcpMessagingTest.java (hunk 14 above)
Comment (samhaa01) [reply coap-core/src/test/java/com/mbed/coap/server/internal/CoapTcpMessagingTest.java]: fixed

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java (hunk 15 above)
Comment (samhaa01) [reply coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java:74]: I think it still tests the socket buffer size. We need it for coverage.

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java (hunk 16 above)
Comment (samhaa01) [reply coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java]: removed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java (hunk 17 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketAdapter.java]: I removed that.

File: coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java (hunk 15 above)
Comment (samhaa01) [reply coap-core/src/test/java/com/mbed/coap/transport/udp/DatagramSocketTransportTest.java]: removed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 13 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: Fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 12 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: Fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java (hunk 2 above)
Comment (samhaa01) [reply coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java]: fixed.

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 18]
```diff
@@ -134,23 +107,25 @@ protected void receive(CoapReceiver coapReceiver, DatagramPacket datagramPacket)
...
-        DatagramSocket datagramSocket = new DatagramSocket(bindSocket);
-        datagramSocket.setSoTimeout(200);
-        return datagramSocket;
+    protected void createSocket() throws SocketException {
+        socket = new DatagramSocketAdapter(bindAddress);
     }
 
     @Override
-    public void stop() {
+    public void stop()  {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:115]: Formatting

File: coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java [hunk 19]
```diff
@@ -160,12 +135,16 @@ public void sendPacket0(CoapPacket coapPacket, InetSocketAddress adr, TransportC
...
     }
 
+    protected boolean socketCreated() {
+        return (socket != null);
+    }
+
     @Override
     public InetSocketAddress getLocalSocketAddress() {
-        return (InetSocketAddress) socket.getLocalSocketAddress();
+        return (getSocket()).getBoundAddress();
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/transport/udp/DatagramSocketTransport.java:144]: extra brackets?

**Review by szysas — APPROVED** [review]
Only minor formatting left.
Check also why code coverage dropped.
